crates/node/src/validate.rs

```diff
@@ -208,6 +279,96 @@ pub async fn validate(
     Ok(ValidateOutcome::Valid(ValidOutcome { total_gas }))
 }
 
+impl DryRun {
+    /// Create a new dry run database which puts the given block into memory
+    /// then cascades from in-memory to on-disk database.
+    pub async fn new(conn_pool: ConnectionPool, block: &Block) -> Result<Self, rusqlite::Error> {
+        let memory = Memory::new(block)?;
+        Ok(Self { memory, conn_pool })
+    }
+}
+
+impl Memory {
+    /// Create a new in-memory database with the given block.
+    fn new(block: &Block) -> Result<Self, rusqlite::Error> {
+        // Only need one connection for the memory database
+        // as there is no contention.
+        let config = db::Config {
+            conn_limit: 1,
+            source: db::Source::Memory(uuid::Uuid::new_v4().to_string()),
+        };
+        let memory = db::ConnectionPool::new(&config)?;
+        let mut conn = memory
+            .try_acquire()
+            .expect("can't fail due to no other connections");
+
+        // Insert and finalize the block.
+        let tx = conn.transaction()?;
+        essential_node_db::create_tables(&tx)?;
+        let hash = essential_node_db::insert_block(&tx, block)?;
+        essential_node_db::finalize_block(&tx, &hash)?;
+        tx.commit()?;
+
+        Ok(Self(memory))
+    }
+}
+
+impl Db {
+    /// Acquire a connection from the database.
+    pub async fn acquire(&self) -> Result<Conn, tokio::sync::AcquireError> {
+        let conn = match self {
+            Db::DryRun(dry_run) => {
+                let cascade = Cascade {
+                    memory: dry_run.memory.as_ref().acquire().await?,
+                    db: dry_run.conn_pool.acquire().await?,
+                };
+                Conn::Cascade(cascade)
+            }
+            Db::ConnectionPool(conn_pool) => Conn::Handle(conn_pool.acquire().await?),
+        };
+        Ok(conn)
+    }
+}
+
+impl Conn {
+    /// Start a transaction.
+    fn transaction(&mut self) -> Result<Transaction<'_>, rusqlite::Error> {
+        match self {
+            Conn::Cascade(cascade) => {
+                let memory = cascade.memory.transaction()?;
+                let db = cascade.db.transaction()?;
+                Ok(Transaction::Cascade(CascadeTransaction { memory, db }))
+            }
+            Conn::Handle(handle) => {
+                let tx = handle.transaction()?;
+                Ok(Transaction::Handle(tx))
+            }
+        }
+    }
+}
+
+/// Cascade from in-memory to on-disk database across transactions.
+fn cascade(
```

Comment:
At first from your PR comment I thought `cascade` was referring to using some sql cascade functionality or something, but I see that you're using it for the query fallback behaviour. I'm not sure about using cascade, but I can see where you're coming from with it.